### ISSUE TITLE
User policies

### ISSUE THREAD

**Issue by hackaugusto** [raiden-network/raiden#216]
Implement a policy class that controls some of the node's behavior:

- Decide what to do once a new channel is openned
  - Ignore a channel from a unknown peer?
  - Automatically add to the available graph and use it?
- Decide how to deposit asset into a channel
  - When a channel is opened, should a deposit be made?
     - A minimum deposit
     - A matching deposit
- If there are no available channels to do a mediated transfer, should a new deposit be made?
  - New channel, to whom?
  - Existing channel, which?
- Control the transfer finalization
  - Should a Secret revealed right away?
  - Should a RequestSecret be sent right away?

An alternative implementation should expose callbacks for these decisions, this can be used to trigger UI changes and request the user's input.

**Comment by LefterisJP** [comment]
@hackaugusto Very good discussion points. Thank you.

@konradkonrad  @heikoheiko and whoever else wants in on the conversation please state your opinion if such user policies should be part of the MVP and if so elaborate on the bullet points presented above.

Some initial input from me.

> Decide what to do once a new channel is opened 
I thought it was already added to the graph. If not, I guess we should do it, I don't see a reason not to.


> Decide how to deposit asset into a channel 

I would assume an initial deposit would need to be made. A channel without a deposit is not very useful, no?

> If there are no available channels to do a mediated transfer, should a new deposit be made?

Here you mean if there is no available path **with a sufficient deposit**? Or if there is no path whatsoever?

> Control the transfer finalization
>    - Should a Secret revealed right away?
>    - Should a RequestSecret be sent right away?

Can you elaborate bit on these points? What is the alternative? It is my understanding that once the recipient of a mediated transfer receives it he has to send the request secret message and then from the other way the secret reveal message would be sent. 


I also like the idea of providing some defaults for all of these policies but allow the user to change them for his node.

**Comment by hackaugusto** [comment]
The current policy is as follows:

- Once a new channel is openned it's automatically added to network graph and used.
- No *automatic* deposit is made
- If there are no available channels to do a mediated transfer the transfer fails
- Secret and RequestSecret messages are sent right away

> I thought it was already added to the graph. If not, I guess we should do it, I don't see a reason not to.

When a new channel is opened a ChannelNew event is created, the current routing implementation keeps a full representation of the existing channels, so yes, currently the default it's to automatically add the new channel into the graph an use it normally.

> I would assume an initial deposit would need to be made. A channel without a deposit is not very useful, no?

Let's say that Alice is openning a channel with Bob, we expect that Alice will open the channel and make it's initial deposit, the policy that I'm referring to is what Bob will do, should it deposit something, match Alice's deposit or do nothing? Currently we do nothing, we assume that Alice is openning a channel with Bob with the intention of making transfers with him, so that would give Bob balance to use for additional transfers.

> Here you mean if there is no available path with a sufficient deposit? Or if there is no path whatsoever?

Both cases, we should probably talk about them independently.

> Can you elaborate bit on these points? What is the alternative? It is my understanding that once the recipient of a mediated transfer receives it he has to send the request secret message and then from the other way the secret reveal message would be sent. 

Added because of the issue #142 .

> I also like the idea of providing some defaults for all of these policies but allow the user to change them for his node.

We do have defaults, but they are hardcoded logic, the issue is to abstract these decision in a Policy class to allow customization.

**Comment by LefterisJP** [comment]
At minimum we will need user policies for:

- Allow customization of the fee function that is sent to PFS
- Specify for which channels to allow monitoring (aka for which channels to send BPs to the MS). That would probably depend on the value held in the channel. That's where the user defined exchange rate oracles mentioned by @konradkonrad  would also come into play.

**Comment by LefterisJP** [comment]
So @Dominik1999 and @rakanalh as we discussed today in the Berlin office. For Ithaca the minimum we would require would be:

1. A configuration value that sets the fee function for the user.
2. As @konradkonrad suggested above a user-configurable Price Oracle. I suppose this can be any of cryptocompare, coinmarketcap, coinpapricka e.t.c.
3. Some user configurable USD value above which channels would have their BPs broadcasted for monitoring.
4. Defaults for all 3 of the above.
5. The above should be made configurable in runtime. We should not require the user to restart raiden to change something.

Then (1) would be used in:
    - Sending the fee function to the PFS at startup and each time the user changes the configuration value.
    - Whenever a transfer is mediated in order to calculate how much to reduce the locked amount for.

(2) and (3) would be used whenever a BP is received for a channel to determine if that BP should be broadcasted or not.

--- 

An alternate approach for these policies would be to make them completely programmatical, callback based. For example a callback `decide_whether_to_broadcast_bp()` for which we provide a default implementation but then user can programmatically load/write their own implementations.

---

Does the following make sense to you guys?

**Comment by Dominik1999** [comment]
Can we aim to close this discussion at the 14.02.? Then we will have two weeks to implement. The policy can change anyway in case we find better parameters.

**Comment by Dominik1999** [comment]
@LefterisJP @rakanalh the User Policy laid out by Lefteris meets our requirements from the PFS / MS side - as discussed with @palango. 

However, from a broader perspective including user experience/user friendliness, we might need to think about it again. As I see it, 
- every user would also need to hold some RDN tokens for the respective UDC contract. RDN is not easy to get at the moment, so maybe we need a function that automatically gets some RDN in exchange for ETH somehow?
- we need many different error messages/notifications, e.g. when there is not enough deposit anymore ...
- we might need an explanation on how to withdraw the tokens from the UDC 
- ... (I think there are many more open questions)

How do we proceed here? Does @heikoheiko want to add something?

**Comment by heikoheiko** [comment]
Yes, whoever is working/planning on setting up user deposits please schedule a call with me.

**Comment by konradkonrad** [comment]
> At minimum we will need user policies for:
> 
>     * Allow customization of the fee function that is sent to PFS
> 
>     * Specify for which channels to allow monitoring (aka for which channels to send BPs to the MS). That would probably depend on the value held in the channel. That's where the user defined exchange rate oracles mentioned by @konradkonrad  would also come into play.

I'm sorry, if my example went too much into detail with the fx oracles -- what I meant to lay out is, that user policies will need to be able to act dynamically. From my line of thinking, I would prefer some plugin mechanism, where we provide an API for such dynamic values, and allow (and encourage) third party integrations. I would rather avoid having e.g. the oracle integration in the client (code/repository).

**Comment by rakanalh** [comment]
For Ithaca, we agree that we'll use pre-defined constant values for behavior-changing settings. Scoping this out of Ithaca until it is required to implement this.
